Switch pane: implement auto toggle switch control

ybott-qinetic opened the pull request:
The AutoToggleSwitch control is a three way switch controlling two states.

/SwitchableOutput/<x>/State sets the default toggle.
 /SwitchableOutput/<x>/Auto sets the additional toggle.

Contributes to #2303

ybott-qinetic commented on components/AutoToggleButton.qml:
```diff
@@ -0,0 +1,190 @@
+                    x: Theme.geometry_tabBar_horizontalMargin
+                    width: parent.width - 2*x
+                    elide: Text.ElideRight
+                    text: "Auto"
```
Replace with CommonWords.auto

ybott-qinetic replied:
Done

blammit commented on components/AutoToggleButton.qml:
```diff
@@ -0,0 +1,190 @@
+        background: Rectangle {
+            anchors.fill: parent
+            radius: Theme.geometry_button_radius
+            color: root.enabled === false ? Theme.color_font_disabled : Theme.color_ok
```
`enabled` is a bool so there's no type coercion involved when testing the condition, so just use `!root.enabled` instead of `root.enabled === false` (here and elsewhere, and also for `checked` usage).

blammit commented on components/AutoToggleButton.qml:
```diff
@@ -0,0 +1,190 @@
+        contentItem: Row {
+            id: buttonRow
+
+            property real buttonWidth: (root.width - (Theme.geometry_button_border_width * 4)) / 3
```
Use theme values instead of constants? Also looks like this property can be `readonly`.

ybott-qinetic replied:
I have added readonly.
I am using Theme values. Each buttons width is the total width of the control minus the borders (left and right as well as two inner borders, which makes four) divided by the total number of buttons that will be displayed in this control, which is fixed at 3. Should these buttons have more Theming in regards to layout? I made this generic so small increases/decreases in width control will not break the layout.

blammit replied:
I see, in that case, add a comment above that explains what the 4 and 3 are for, or alternatively a `readonly property int buttonCount: 3` and then use that property.

blammit commented on components/AutoToggleButton.qml:
```diff
@@ -0,0 +1,190 @@
+                    anchors.fill: parent
+                    onPressed: {
+                        if (root.autoChecked === false) {
+                            root.toggled()
```
It shouldn't toggle between on and off; it should only select the button that you have clicked. Otherwise it's confusing if you select "off" but it changes to "on".

ybott-qinetic replied:
Done

blammit commented on components/AutoToggleButton.qml:
```diff
@@ -0,0 +1,190 @@
+                    width: parent.width - 2*x
+                    elide: Text.ElideRight
+                    text: CommonWords.on
+                    color: root.enabled === false
```
To avoid the code duplication between the three buttons, use an inline component like:

```
component InternalButton : Rectangle {
    property alias text: buttonLabel.text
    
    Label { id: buttonLabel }
    MouseArea { ... }
}
```

and then your content item can just be:

```
contentItem: Row {
    InternalButton {
        text: "Off"
        bottomLeftRadius: Theme.geometry_button_radius - Theme.geometry_button_border_width
        topLeftRadius: Theme.geometry_button_radius - Theme.geometry_button_border_width
    }
    InternalButton {
        text: "Off"
    }
    InternalButton {
        text: "Auto"
        bottomRightRadius: Theme.geometry_button_radius - Theme.geometry_button_border_width
        topRightRadius: Theme.geometry_button_radius - Theme.geometry_button_border_width        
    }
}
```


But also, would it be easier to just use `Button` instead of a custom component? That comes with a built-in press-effect, and that is missing from this current implementation. It also comes with the key navigation handling.

ybott-qinetic replied:
Done.

blammit commented on components/AutoToggleButton.qml:
```diff
@@ -0,0 +1,190 @@
+        switch (key) {
+        case Qt.Key_Space:
+            if (activeFocus) {
+                offButton.focus = true
```
Move this `handlePress()` into the `SwitchableOutputDelegate`, as it's only used from that context.

And then it should be sufficient to `focus=true` on this whole AutoToggleButton control, rather than a specific button.

Currently, once you press Space and the focus is taken by this control, the focus cannot be removed, so you need a `Keys.onEnterPressed` and `Keys.onReturnPressed` handler here to remove the focus from the control.

blammit commented on components/AutoToggleButton.qml:
```diff
@@ -0,0 +1,190 @@
+FocusScope {
+    id: root
+
+    property int fontPixelSize: Theme.font_size_body3
```
Looks like the indentation in this file is using spaces rather than tabs.

blammit commented on components/SwitchableOutputDelegate.qml:
```diff
@@ -498,6 +498,49 @@ BaseListItem {
+
+			KeyNavigationHighlight.active: autoToggleButton.activeFocus
+
+			VeQuickItem {
```
Use a `SettingSync` instead of writing directly to a `VeQuickItem` here for the auto state. And then actually, the "auto" button should be enabled depending on whether its sync state is busy. And instead of enabling/disabling the entire control based on the `toggleState`, just enable the on/off buttons based on the toggle state, and the auto button based on the auto state.

blammit commented on components/SwitchableOutputDelegate.qml:
```diff
@@ -498,6 +498,49 @@ BaseListItem {
+			onOffClicked: toggleState.writeValue(0)
+			onAutoClicked: autoState.setValue(autoState.value === 1 ? 0 : 1)
+
+			KeyNavigationHighlight.active: autoToggleButton.activeFocus
```
This is not needed as we don't want to focus the entire control, just the individual controllable buttons.

ybott-qinetic replied:
Removed.

blammit commented on components/SwitchableOutputDelegate.qml:
```diff
@@ -498,6 +498,49 @@ BaseListItem {
+			fontPixelSize: Theme.font_size_body1
+			enabled: !toggleState.busy
+			checked: toggleState.backendValue === 1 ? true : false
+			autoChecked: autoState.value === 1 ? true : false
```
Don't need the `true` and `false`?

ybott-qinetic replied:
Removed.

ybott-qinetic commented on components/AutoToggleButton.qml:
```diff
@@ -0,0 +1,144 @@
+			InternalButton {
+				id: offButton
+
+				onPressed: if (!root.autoChecked) {
```
Fixed.

ybott-qinetic commented on components/AutoToggleButton.qml:
```diff
@@ -0,0 +1,144 @@
+		KeyNavigationHighlight.active: activeFocus
+	}
+
+	Keys.onSpacePressed: offButton.focus = true
```
I think this is fixed with your suggestions.

ybott-qinetic commented on components/AutoToggleButton.qml:
```diff
@@ -0,0 +1,144 @@
+			InternalButton {
+				id: onButton
+
+				onPressed: if (!root.autoChecked) {
```
Waiting for disabled design feedback.

blammit replied:
Actually, on second thoughts, I think it should be this way instead: the button should still be clickable when autoChecked=true, but, if you click on/off in this state, then the Auto button becomes deselected. (And obviously, handle this behaviour from SwitchableOutputDelegate.qml rather than from within AutoToggleButton.qml.) Because if you click on/off, you are essentially removing the automatic behaviour.

ybott committed:
Switch pane: implement auto toggle switch control

The AutoToggleSwitch control is a three way switch controlling
two states.

/SwitchableOutput/<x>/State sets the default toggle.
/SwitchableOutput/<x>/Auto sets the additional toggle.

Contributes to #2303

ybott-qinetic commented on components/AutoToggleButton.qml:
```diff
@@ -0,0 +1,139 @@
+FocusScope {
+	id: root
+
+	property int fontPixelSize: Theme.font_size_body3
```
Done

blammit commented on cmake/ModuleVenus_Sources.cmake:
```diff
@@ -23,6 +23,7 @@ set (VictronVenusOS_QML_MODULE_SOURCES
     components/ArcGauge.qml
     components/ArcGaugeQuantityRow.qml
     components/AsymmetricRoundedRectangle.qml
+    components/AutoToggleButton.qml
```
Can you put this in `components/controls` instead?

ybott-qinetic replied:
Done

blammit commented on components/AutoToggleButton.qml:
```diff
@@ -0,0 +1,111 @@
+		Keys.onUpPressed: {}
+		Keys.onDownPressed: {}
+		Keys.onLeftPressed: {}
+		Keys.onRightPressed: {}
```
As discussed, these handlers need to be in the base item

ybott-qinetic commented on src/baseautotoggleswitch.h:
```diff
@@ -0,0 +1,42 @@
+	Q_PROPERTY(int buttonCount READ buttonCount CONSTANT FINAL)
+
+public:
+	BaseAutoToggleSwitch();
```
Done

ybott-qinetic commented on src/baseautotoggleswitch.h:
```diff
@@ -0,0 +1,42 @@
+	bool m_onChecked = false;
+	bool m_autoChecked = false;
+
+	int m_buttonCount = 3;
```
Done

blammit commented:
@chriadam could you check this one as well? The control now has a template implemented in C++, as discussed.

chriadam commented on components/SwitchableOutputDelegate.qml:
```diff
@@ -566,6 +566,49 @@ BaseListItem {
+			onChecked: toggleState.backendValue === 1
+			autoChecked: autoToggleState.backendValue
+			onOnClicked: {
+				autoToggleState.writeValue(0)
```
I think we should disallow clicks to on/off if auto is currently checked, but maybe I'm wrong?

blammit replied:
I wasn't sure at the time, but Rein just confirmed in the latest design update that we should disallow this, yes. If Auto is checked, then show a toast notification if the user tries to click on/off.

ybott-qinetic replied:
I have implemented this as well as the toast notification.

chriadam commented on src/baseautotoggleswitch.cpp:
```diff
@@ -0,0 +1,38 @@
+BaseAutoToggleSwitch::BaseAutoToggleSwitch(QQuickItem *parent)
+    : QQuickControl(parent)
+{
+	setFlag(ItemIsFocusScope);
```
might also need to set TabFocus policy but not sure - check with Bea

ybott-qinetic replied:
I have set the policy for this.
